src/wrappers/themis/python/pythemis/scell.py

```diff
@@ -24,11 +24,42 @@
 themis = cdll.LoadLibrary(find_library('themis'))
 
 
+class SecureCellError(ThemisError):
+    def __init__(self, message, error_code=THEMIS_CODES.INVALID_PARAMETER):
+        message = 'Secure Cell: ' + message
+        super(SecureCellError, self).__init__(error_code, message)
+
+
 class SCellSeal(object):
+    def __new__(cls, key=None, passphrase=None):
```

Comment:
> why we need to use `__new__` instead of `__init__`?

`__new__` is used to have `SCellSeal(passphrase='smth')` return an instance of `SCellSealPassphrase` as `__new__` can choose to construct an instance of different class.

With `__init__` alone it is not possible to make `SCellSeal(passphrase='smth')` behave like `SCellSealPassphrase('smth')`.

Arguably, we could remove `__new__` and have `SCellSealPassphrase('smth')` as the only way to initialize Secure Cell with a passphrase, but I think that using a single `SCellSeal` is more appropriate since the only difference in API is initialization.

Reply:
hm, from this point looks good. only one note, that we can have same problem when people may think that default behavior of SCellSeal is to expect passphrase instead key. But it expects key ) it is a problem why we started providing passphrase api ) 
but we want to save backward compatibility, so we can't force to explicitly define which of keyword parameter is used, so I agree with your version.

Reply:
Yeah, `SCellSeal(b'password')` will treat it as a key, just like it did before. This behavior stays here for compatibility.

However, now it does more strict type checks so `SCellSeal('password')` will raise an exception from now on. PyThemis 0.12 allows that expression, considering it a key (by encoding the string in some implementation-defined encoding).

It could be possible to dispatch solely on the type – that is, have `SCellSeal(b'password')` treat it as a key but `SCellSeal('password')` to use it as a passphrase – but that's _way_ too unexpected behavior. It also silently alters behavior depending on whether the code is executed by Python 2 or Python 3. So I'd rather make it explicit with keyword arguments.

Reply:
hm... I missed that now we will not allow using SCellSeal('password') ...
I'm not sure that we should remove it because someone can use it.
Heh, I just looked at how we use `authentication token` in our [simulator](https://docs.cossacklabs.com/simulator/data-cell/) and checked that we encode it to bytes before using SCellSeal ))
But I admit that exists someone who pass strings to SCellSeal and use strong long strings as key ))

@vixentael what you think, should we remove SCellSeal constructor that allow to use string as key or only bytes and if someone wants to use string, he should use passphrase='' parameter?

Reply:
After thinking about this for a while I believe that it was too rash to raise exceptions in this case. While it would have been appropriate for a _new API_ (as it was originally planned), if we reuse the old API then “Thou Shalt Not Break Existing Code” rule is in effect.

There may be some users that legitimately write `SCellSeal(passphrase)` where passphrase is actually a 100-character long passphrase, as a Unicode string. There also might be Python 2 users that write `SCellSeal('passphrase')` and we treat that as valid master key because it's `str`.

However, I still feel that we should tell the users when they do something strange that was permitted before but is no longer recommended. Raising an exception is pretty effective way at carrying over your message, but it's a breaking change after all.

Instead, I have replaced all new exceptions raised from existing code with _warnings_. They will still be visible to people who care about them, but they won't break existing code with default settings.

@Lagovas, @vixentael, is this an acceptable compromise?

Reply:
warnings are a good compromise, imho